### PR TITLE
Pricing backend

### PR DESCRIPTION
**Pull request by alneberg**
Work in progress for refactoring pricing backend.

  - [x] Have a clear separation between data fetching api in python, routing of pages in python and UX-interface in js.
  - [x] Merge the two databases pricing_components and pricing_products into a single cost_calculator
  - [x] Have pricing_update fetch all data through the api
  - [x] Have pricing_quote work off the new cost_calculator database
  - [x] Implement save action from pricing_update
  - [x] Implement pricing preview page where a draft can be created
  - [x] Pricing preview page where a list of changes in current draft can be seen
  - [x] Validation api that can return validation errors and also list all the changes compared to the current published
  - [x] Error handling for all the api fetching methods

Pricing quote specifically:
 - [x] Load data through Vue and use the cost_calculator database
 - [x] Possible to create quote with Vue
 - [x] Possible to show discontinued products (make sure datatables work)

### CODE REVIEW

File: status/pricing.py [hunk 1]
```diff
@@ -20,379 +20,67 @@ def _validate_version_param(self, version):
...
-            if not discontinued and product['Status'] != 'Available':
-                return_d.pop(product_id)
-                continue
+    def _update_last_modified(self, doc):
+        user_name = self.get_current_user().name
+        user_email = self.get_current_user().email
 
-            price_int, price_acad, price_full = self._calculate_product_price(product, all_component_prices)
+        doc['Last modified by user'] = user_name
+        doc['Last modified by user email'] = user_email
```
Comment (aanil) [inline status/pricing.py:61]: Would it be better to have user and user email as a tuple?

File: status/pricing.py [hunk 2]
```diff
@@ -409,80 +97,49 @@ class PricingExchangeRatesDataHandler(PricingBaseHandler):
...
-                                               pretty_strings=True)
-
-        t = self.application.loader.load('pricing_quote_tbody.html')
-        self.write(t.generate(gs_globals=self.application.gs_globals,
-                              user=self.get_current_user(),
-                              products=products,
-                              components=components,
-                              version=version))
 
+        if date is not None:
```
Comment (aanil) [inline status/pricing.py:133]: Do we need this check? `date `is set to `None` above and `_fetch_exchange_rates` handles that case anyway

File: status/pricing.py [hunk 3]
```diff
@@ -583,109 +255,340 @@ def _validate_discontinued(self, components, products):
...
-        except Exception:
-            raise # Maybe this is the default?
+        """ Should return the specified version of the cost calculator."""
+        doc = self.fetch_latest_doc()
+        draft = doc['Draft']
+
+        response = dict()
+        if draft:
+            response['cost_calculator'] = doc
+        if not draft:
```
Comment (aanil) [inline status/pricing.py:367]: Is this check needed? There could be a default `response['cost_calculator'] = None` before `if draft:`

**Review by pericsson — APPROVED** [review]

File: status/pricing.py [hunk 4]
```diff
@@ -20,379 +20,67 @@ def _validate_version_param(self, version):
...
 
-        sek_price = sek_list_price - sek_list_price * component['Discount']
+    def fetch_latest_doc(self):
+        db = self.application.cost_calculator_db
+        curr_rows = db.view('entire_document/by_version', descending=True, limit=1).rows
+        curr_doc = curr_rows[0].value
+        return curr_doc
 
-        sek_price_per_unit = sek_price/float(component['Units'])
+    def fetch_latest_published(self):
```
[the remark this answers is not in the document]
Comment (alneberg) [reply status/pricing.py]: Done!

File: status/pricing.py (hunk 1 above)
Comment (alneberg) [reply status/pricing.py:59]: A tuple in the json doc? Or do you mean similar to the lock_info where there is a two level thing, so for example: 

```
doc['Last modified'] = {'user': user_name, 'email': user_email}
```
?

I would agree that that would be better, but I'm not sure it's big enough of an improvement to be worth the change?

File: status/pricing.py (hunk 2 above)
Comment (alneberg) [reply status/pricing.py]: You're absolutely right. Fixed!

File: status/pricing.py (hunk 3 above)
Comment (alneberg) [reply status/pricing.py]: Fixed!

File: run_dir/static/js/pricing_main.js [hunk 5]
```diff
@@ -0,0 +1,998 @@
...
+                    }
+                    prod_per_cat[cat][product['REF_ID']] = product
+                }
+            }
+            return prod_per_cat
+        },
+        product_categories() {
+            categories = new Set();
+            for ([prod_id, product] of Object.entries(this.all_products)) {
+                cat = product['Category']
```
[the remark this answers is not in the document]
Comment (alneberg) [reply run_dir/static/js/pricing_main.js]: Good catch! Removed all four of them now.

File: run_dir/static/js/pricing_main.js [hunk 6]
```diff
@@ -0,0 +1,994 @@
...
+              <div class="modal-header">
+                <h4 class="modal-title" id="myModalLabel">Fetch historic exchange rate</h4>
+                <button type="button" class="btn-close" data-dismiss="modal" aria-label="Close"></button>
+              </div>
+              <div class="modal-body">
+                <p>The most recent rates prior to the date chosen will be used.</p>
+                <p>Latest date exchange rates: <input id="datepicker" type='date' v-model="issued_at_form_bound" data-date-format="yyyy-mm-dd"></p>
+              </div>
+              <div class="modal-footer">
+                <button type="button" class="btn btn-default" data-dismiss="modal">Cancel</button>
```
Comment (aanil) [inline run_dir/static/js/pricing_main.js:493]: btn-default has no styling in bootstrap 5

File: run_dir/static/js/pricing_update.js [hunk 7]
```diff
@@ -816,104 +715,104 @@ app.component('components', {
...
       },
-      cost() {
-          return this.$root.componentCost(this.component_id)
+      removeComponent(event) {
+          /* This took me some time to figure out - the problem is that the bootstrap tooltip would
+          remain open forever if it's not actively removed like this */
+          if (event) {
+              tooltip = bootstrap.Tooltip.getInstance(this.$el)
+              tooltip_el = this.$el
+              that = this
```
Comment (aanil) [inline run_dir/static/js/pricing_update.js:784]: 😄

File: run_dir/static/js/pricing_update.js [hunk 8]
```diff
@@ -816,104 +715,104 @@ app.component('components', {
...
+      <tr data-toggle="tooltip" data-placement="top" :data-original-title="tooltip_html" data-animation=false data-html=true>
+        <th scope="row">{{component_id}}</th>
+        <td>{{component['Product name']}}</td>
+        <td v-if="added">
+          <!-- I was for some reason unable to solve this with regular v-model... -->
+          <input class="form-control" type="number" min=1 :value="component_data['quantity']" @input="updateQuantity($event.target.value)"/>
+        </td>
+        <td class="pl-3 pt-2" v-if="added">
+          <a class="mr-2" href="#" @click="this.removeComponent">
+            <i class="far fa-times-square fa-lg text-danger"></i>
```
Comment (aanil) [inline run_dir/static/js/pricing_update.js:805]: should the class be `fa` instead of `far`?

File: run_dir/static/js/pricing_update.js [hunk 9]
```diff
@@ -816,104 +715,104 @@ app.component('components', {
...
+          </a>
+        </td>
+        <td class="pl-3 pt-2" v-if="!added">
+          <a class="mr-2" href="#" @click="this.addComponent">
+            <i class="far fa-plus-square fa-lg text-success"></i>
```
Comment (aanil) [inline run_dir/static/js/pricing_update.js:810]: Here too

File: status/pricing.py (hunk 1 above)
Comment (aanil) [reply status/pricing.py:59]: The draft document is huge anyway, so I guess a few more lines won't make a difference.

File: run_dir/static/js/pricing_main.js (hunk 6 above)
Comment (alneberg) [reply run_dir/static/js/pricing_main.js]: Fixed!

File: run_dir/static/js/pricing_update.js (hunk 8 above)
Comment (alneberg) [reply run_dir/static/js/pricing_update.js:805]: Nope

File: run_dir/static/js/pricing_update.js (hunk 9 above)
Comment (alneberg) [reply run_dir/static/js/pricing_update.js:810]: No here too. 😆

**Review by aanil — APPROVED** [review]
Everything looks ok to me. Absolutely impressive job!

**Comment by aanil** [comment]
😱 😱

**Comment by alneberg** [comment]
💥

**Pull request merged**